### PR TITLE
[asl] Implement load-reserve store conditional

### PR DESCRIPTION
**Pull request by maranget**
To complete PR #645.

### CODE REVIEW

File: asllib/Interpreter.mli [hunk 1]
```diff
@@ -22,6 +22,7 @@
 module type S = sig
   module B : Backend.S
 
+  val run_env : (AST.identifier * B.value) list -> B.ast -> B.value B.m
```
Comment (HadrienRenaud) [inline asllib/Interpreter.mli:25]: Can we put some documentation here? The name is not really explicit

File: herd/ASLSem.ml [hunk 2]
```diff
@@ -385,6 +392,19 @@ module Make (C : Config) = struct
...
+    let checkprop m_prop =
+      let* vprop = m_prop in
+      M.assign (vbool true) vprop >>! []
+
+    let somebool _ =
+      let v = V.fresh_var () in
+      let mbool b =
+        let vb = V.Val (Constant.Concrete (ASLScalar.S_Bool b)) in
+        M.assign v vb >>= fun () -> M.unitT vb  in
+      choice (M.unitT v) (mbool true) (mbool false)
```
Comment (HadrienRenaud) [inline herd/ASLSem.ml:406]: Is it exactly what the ASL syntax `UNKNOWN: boolean` does?

File: herd/ASLSem.ml [hunk 3]
```diff
@@ -385,6 +392,19 @@ module Make (C : Config) = struct
     (* Primitives and helpers                                                 *)
     (**************************************************************************)
 
+    let vbool b =  V.Val (Constant.Concrete (ASLScalar.S_Bool b))
+
+    let checkprop m_prop =
+      let* vprop = m_prop in
+      M.assign (vbool true) vprop >>! []
```
Comment (HadrienRenaud) [inline herd/ASLSem.ml:399]: Is that different from the ASL syntax `assert prop`?

File: herd/libdir/asl-pseudocode/implementations.asl [hunk 4]
```diff
@@ -1,3 +1,111 @@
...
+var SuccessVA : boolean ;
+
+func AArch64_IsExclusiveVA
+(address :: bits(64), processorid :: integer, size :: integer) => boolean
+begin
+  // Try both possibilties: write or not write
+  SuccessVA = SomeBoolean();
+  let reserved = RESADDR;
+  // If write succeeds then effective address and reservation coincide.
+  if SuccessVA then CheckProp(address == reserved); end
```
Comment (HadrienRenaud) [inline herd/libdir/asl-pseudocode/implementations.asl:61]:
```suggestion
  if SuccessVA then assert address == reserved; end
```

File: asllib/Interpreter.ml [hunk 5]
```diff
@@ -1009,16 +1017,21 @@ module Make (B : Backend.S) (C : Config) = struct
         in
         Error.fatal_unknown_pos (Error.UncaughtException msg)
 
+  let run_typed ast env = run_typed_env [] ast env
+
   (** Main entry point for the Interpreter, [run ast primitives] type-annotate
       [ast], build a global environment and then evaluate the "main" function
       in it. *)
```
Comment (HadrienRenaud) [inline asllib/Interpreter.ml:1024]: I think we can delete those at this point

File: herd/AArch64ASLSem.ml [hunk 6]
```diff
@@ -445,62 +473,30 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
...
 
-    let symb_values_tbl : (string, V.v) Hashtbl.t = Hashtbl.create 17
-    let symb_values_tag = "AArch64-HashedValue"
+    let tr_cst tr =
+      Constant.map tr
+        (fun _ -> Warn.fatal "Cannot translate PTE")
+        (fun _ -> Warn.fatal "Cannot translate instruction")
+
+    let aarch64_to_asl_bv = function
+      | V.Var v -> ASLS.A.V.Var v
```
Comment (HadrienRenaud) [inline herd/AArch64ASLSem.ml:482]: Will there be a problem if the symb counter on ASL values is not increased and we have symbolic values with the same number?
I think we should use fresh_var and keep track of why ASL value equals which AArch64 value, or make the counter of symbolic values common for all value types.

File: herd/AArch64ASLSem.ml [hunk 7]
```diff
@@ -445,62 +473,30 @@ module Make (TopConf : AArch64Sig.Config) (V : Value.AArch64ASL) :
...
-          (* The initial value for PSTATE given bellow
-           * overrides the one of the "patches.asl" file,
-           *  see `herd/ASLSem.ml`.
-           *)
-          ::(MiscParser.Location_reg (proc,pstate),tc)
-          ::init_gregs
-        | None -> init_gregs
-        in
-        let prog =
+      let init = [] in
```
Comment (HadrienRenaud) [inline herd/AArch64ASLSem.ml:498]: Why do we still have this?
And why did you move the construction of init after?

File: asllib/Interpreter.mli (hunk 1 above)
Comment (maranget) [reply asllib/Interpreter.mli:25]: Of course, will do.

File: herd/libdir/asl-pseudocode/implementations.asl [hunk 8]
```diff
@@ -1,3 +1,111 @@
...
+//
+// It is always safe to return TRUE which will check the physical address only.
+
+var SuccessVA : boolean ;
+
+func AArch64_IsExclusiveVA
+(address :: bits(64), processorid :: integer, size :: integer) => boolean
+begin
+  // Try both possibilties: write or not write
+  SuccessVA = SomeBoolean();
```
[the remark this answers is not in the document]
Comment (maranget) [reply herd/libdir/asl-pseudocode/implementations.asl:58]: I was not aware of  UNKOWN. `SomeBoolean()` returns an alternative monad `M.altT`. with one alternative returning true and the other returning false. I have to check that `UNKNOWN` does the same...

File: herd/ASLSem.ml (hunk 2 above)
Comment (maranget) [reply herd/ASLSem.ml:406]: No idea!

File: herd/AArch64ASLSem.ml (hunk 6 above)
Comment (maranget) [reply herd/AArch64ASLSem.ml:482]: Yes it can be a problem if two different variables are the same.  This can be the case, I am afraid.

File: herd/AArch64ASLSem.ml (hunk 7 above)
Comment (maranget) [reply herd/AArch64ASLSem.ml:498]: Aha, If we construct init before calling `ASLTH.build`  we  have to turn AArch64ASL values into `ParsedConstant.v`'s. When we do it after, we translate AArch64ASL values into ASL values, which are basically the same!

File: herd/ASLSem.ml (hunk 3 above)
Comment (maranget) [reply herd/ASLSem.ml:399]: I have not checked, but I assume this is different: `checkprop` produces an equation that will be contradictory if vprop is `FALSE`; while `ASSERT` raises an exception that may have less controlled consequences.

File: asllib/Interpreter.ml (hunk 5 above)
Comment (maranget) [reply asllib/Interpreter.ml:1024]: Delete comment?

File: herd/AArch64ASLSem.ml (hunk 6 above)
Comment (maranget) [reply herd/AArch64ASLSem.ml:482]: A new commit 5d08b1cc82e379a11a717cacc7f0bb131f2d95db intend to fix this potential problem.

File: herd/AArch64ASLSem.ml (hunk 7 above)
Comment (HadrienRenaud) [reply herd/AArch64ASLSem.ml:498]: Oh smart!

File: herd/ASLSem.ml (hunk 2 above)
Comment (HadrienRenaud) [reply herd/ASLSem.ml:406]: After checking, in ASLSem, for the moment, `UNKNOWN: _t` produces a fresh variable. As you have a condition just in front of it, it will create an altT with the two possibilities.

Comment (maranget) [reply herd/ASLSem.ml:406]: Do you mean that we can write `SomeBoolean()` using `UNKNOW`? Something like
```
func SomeBoolean() => boolean
begin
  return if UNKNOWN:boolean then TRUE else FALSE end;
end
```
(Have not tried).

Comment (HadrienRenaud) [reply herd/ASLSem.ml:406]: This would indeed have the same effect as writing it as a primitive.

But actually, what I was suggesting is not to implement any function `SomeBool` and just use `UNKNOWN: boolean` instead of calling the function.

File: herd/ASLSem.ml (hunk 3 above)
Comment (HadrienRenaud) [reply herd/ASLSem.ml:399]: Maybe we should implement assertions like this then?

File: asllib/Interpreter.ml (hunk 5 above)
Comment (HadrienRenaud) [reply asllib/Interpreter.ml:1024]: Yes: they are not alongside their function as you put run_env between them, and do not have the correct arguments anymore.

File: herd/ASLSem.ml (hunk 3 above)
Comment (maranget) [reply herd/ASLSem.ml:399]: I am not sure we should. Assertions are here to check for errors. We probably want errors to be flagged as such and not to disappear silently.

File: herd/ASLSem.ml (hunk 2 above)
Comment (maranget) [reply herd/ASLSem.ml:406]: This is doable for `UNKNOWN:boolean` but not for `UNKNOWN:integer`. So we would have a significant discrepancy of semantics. This would be disturbing for users, or?

File: herd/ASLSem.ml (hunk 3 above)
Comment (HadrienRenaud) [reply herd/ASLSem.ml:399]: Okay then.

File: herd/ASLSem.ml (hunk 2 above)
Comment (HadrienRenaud) [reply herd/ASLSem.ml:406]: I don't understand the parallel with integers, we are not implementing `SomeInteger` here?

For clarification I am not suggesting to change the semantics of `UNKNOWN: boolean` to match those of `SomeBoolean`.

Comment (maranget) [reply herd/ASLSem.ml:406]: I have tried to replace the call to `SomeBoolean()` by:
```
SuccessVA =
    if UNKNOWN:boolean then TRUE else FALSE;
```
It does not work. Lacking time to experiment, I keep the current solution with a specific `SomeBoolean()` primitive. Please do not hesitate to post a follwoup, even after merging, which is imminent.

**Comment by maranget** [comment]
Hi @HadrienRenaud I have merged the PR. Thanks for your review. I could see how to use the existing `UNKNOWN::boolean` constructing in place of the new primitive `SomeBoolean()`. Please elaborate if you see how to proceed. Thanks again.

File: herd/ASLSem.ml (hunk 2 above)
Comment (HadrienRenaud) [reply herd/ASLSem.ml:406]: Okay, a shame but not too costly. It'd be good to investigate why using `UNKNOWN:boolean` does not work there.
